### PR TITLE
Split mapreg table and refactor mapreg code

### PR DESCRIPTION
**Pull request by Kenpachi2k13**
### Pull Request Prelude

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

Split `mapreg` SQL table into separate tables for integer and string variables as a first step for the planned script engine update.
Additionally refactor code in *src/map/mapreg_sql.c* and *src/map/mapreg.h*.

**Issues addressed:** None.

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: src/map/mapreg_sql.c [hunk 1]
```diff
@@ -105,6 +105,25 @@ static bool mapreg_setreg(int64 uid, int val)
...
 				if( SQL_ERROR == SQL->Query(map->mysql_handle, "INSERT INTO `%s`(`varname`,`index`,`value`) VALUES ('%s','%u','%d')", mapreg->table, tmp_str, i, val) )
 					Sql_ShowDebug(map->mysql_handle);
+
+				struct SqlStmt *stmt = SQL->StmtMalloc(map->mysql_handle);
+
+				if (stmt == NULL) {
+					SqlStmt_ShowDebug(stmt);
+				} else {
+					const char *query = "INSERT INTO `%s` (`key`, `index`, `value`) VALUES (?, ?, ?)";
+					size_t len = strnlen(name, SCRIPT_VARNAME_LENGTH);
```
Comment (4144) [inline src/map/mapreg_sql.c]: better use ``safestrnlen``

File: src/map/mapreg_sql.c [hunk 2]
```diff
@@ -240,9 +417,83 @@ static void script_load_mapreg(void)
...
+ *
+ * @param name The variable's name.
+ * @param index The variable's array index.
+ * @param value The variable's value.
+ *
+ **/
+static void mapreg_save_num_db(const char *name, unsigned int index, int value)
+{
+	nullpo_retv(name);
+	Assert_retv(strlen(name) > 0);
```
Comment (4144) [inline src/map/mapreg_sql.c:482]: probably faster to check ``*name != 0``

File: src/map/mapreg.h [hunk 3]
```diff
@@ -29,6 +29,8 @@
 struct config_setting_t; // common/conf.h
 struct eri;
 
+#define MAPREG_AUTOSAVE_INTERVAL (300 * 1000)
```
Comment (4144) [inline src/map/mapreg.h]: better change to
```
#ifndef MAPREG_AUTOSAVE_INTERVAL
#define MAPREG_AUTOSAVE_INTERVAL (300 * 1000)
#endif
```

File: src/map/mapreg_sql.c [hunk 4]
```diff
@@ -539,80 +540,92 @@ static void mapreg_save_str_db(const char *name, unsigned int index, const char
...
+ * @param data The DB data holding the mapreg_save data.
+ * @param ap Unused.
+ * @return 0 on success, otherwise 1.
+ *
+ **/
+static int mapreg_destroy_reg(union DBKey key, struct DBData *data, va_list ap)
 {
-	struct mapreg_save *m = NULL;
-
 	if (data->type != DB_DATA_PTR) // Sanity check
```
Comment (4144) [inline src/map/mapreg_sql.c:602]: before this need nullpo check for data

File: src/map/mapreg_sql.c (hunk 1 above)
Comment (Kenpachi2k13) [reply src/map/mapreg_sql.c]: `strnlen()` isn't used anymore in later commits.

File: src/map/mapreg_sql.c (hunk 2 above)
Comment (Kenpachi2k13) [reply src/map/mapreg_sql.c]: Done.

File: src/map/mapreg.h (hunk 3 above)
Comment (Kenpachi2k13) [reply src/map/mapreg.h]: Done.

File: src/map/mapreg_sql.c (hunk 4 above)
Comment (Kenpachi2k13) [reply src/map/mapreg_sql.c:604]: Done.

**Pull request merged**